NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/clminfo2.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "clm info2" 1 "16 May 2014" "clm info2 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
clm info2 \- compute performance measures for graphs and clusterings\&.

clminfo2 is not in actual fact a program\&. This manual
page documents the behaviour and options of the clm program when
invoked in mode \fIinfo2\fP\&. The options \fB-h\fP, \fB--apropos\fP,
\fB--version\fP, \fB-set\fP, \fB--nop\fP are accessible
in all \fBclm\fP modes\&. They are described
in the \fBclm\fP manual page\&.
.SH SYNOPSIS

\fBclm info2\fP [options] <graph file> <cluster file> <cluster file>*

\fBclm info2\fP
\fB[-o\fP fname (\fIwrite to file \fBfname\fP\fP)\fB]\fP
\fB[-pi\fP f (\fIapply inflation beforehand\fP)\fB]\fP
\fB[--list\fP (\fIlist efficiency for all nodes\fP)\fB]\fP
\fB[-tf\fP spec (\fIapply tf-spec to input matrix\fP)\fB]\fP
\fB[-cl-ceil\fP <num> (\fIskip clusters of size exceeding \fB<num>\fP\fP)\fB]\fP
\fB[-cat-max\fP <num> (\fIdo at most \fB<num>\fP tree levels\fP)\fB]\fP
\fB[-cl-tree\fP fname (\fIexpect file with nested clusterings\fP)\fB]\fP
\fB[-t\fP <int> (\fIuse <int> threads\fP)\fB]\fP
\fB[-J\fP <intJ> (\fIa total of <intJ> jobs are used\fP)\fB]\fP
\fB[-j\fP <intj> (\fIthis job has index <intj>\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
<matrix file> <cluster file> <cluster file>*
.SH DESCRIPTION

\fBclm info2\fP is a streamlined and updated version of \fBclm info\fP\&. The
latter outputs a key-value format listing a number of measures\&. In contrast,
\fBclm info2\fP only outputs the so-called efficiency criterion, a quality
index for networks and clusterings\&. This criterion can be generated for
each node independently with the \fB--list\fP option, indicating how
well a clustering captures the neighbour distribution of a given node\&.

\fBclm info2\fP can utilise threading and job dispatching\&. This may be useful
when dealing with very large graphs\&.

Multiple clusterings can be supplied on the command-line\&.
Output is tabular, each row corresponding with a clustering in the
ordering as supplied on the command line\&. Multiple columns will
result only if node-wise output is induced with \fB--list\fP\&.
By default a single number is produced for each individual clustering:
the mean of all node-wise scores for that clustering\&.

The \fBefficiency\fP factor is described in [1] (see
the \fBREFERENCES\fP section)\&. It tries to balance the dual aims of
capturing a lot of edges or edge weights and keeping the cluster footprint
or area fraction small\&. The efficiency number has several appealing
mathematical properties, cf\&. [1]\&.
.SH OPTIONS

.ZI 2m "\fB-o\fP fname (\fIoutput file name\fP)"
\&
.br
.in -2m

.ZI 2m "\fB-pi\fP f (\fIapply inflation beforehand\fP)"
\&
.br
Apply inflation to the graph matrix and compute the performance
measures for the result\&.
.in -2m

.ZI 2m "\fB-tf\fP <tf-spec> (\fItransform input matrix values\fP)"
\&
.br
shared_defopt{-tf}
.in -2m

.ZI 2m "\fB--list\fP (\fIlist efficiency for all nodes\fP)"
\&
.br
The efficiency scores for all nodes are given on a single line\&.
Each clustering specified corresponds to a single line\&.
.in -2m

.ZI 2m "\fB-cl-tree\fP fname (\fIexpect file with nested clusterings (cone format)\fP)"
\&
'in -2m
.ZI 2m "\fB-cl-ceil\fP <num> (\fIskip (nested) clusters of size exceeding <num>\fP)"
\&
'in -2m
'in +2m
\&
.br
The specified file should contain a hierarchy of nested
clusterings such as generated by \fBmclcm\fP\&. The output is then
in a special format, undocumented but easy to understand\&.
Its purpose is to help cherrypick a single clustering
from a tree, in conjunction with the slightly experimental
and undocumented program \fBmlmfifofum\fP\&.

The measure that is used is very slow to compute for large clusters, and
generally it will be outside any interesting range (i\&.e\&. it will be small)\&.
Use \fB-cl-ceil\fP to skip clusters exceeding the specified size \-
\fBclm info\fP will directly proceed to subclusters if they exist\&.
.in -2m

.ZI 2m "\fB-cat-max\fP num (\fIdo at most num levels\fP)"
\&
.br
This only has effect when used with \fB-cl-tree\fP\&.
\fBclm info\fP will start at the most fine-grained level, working upwards\&.
.in -2m

.ZI 2m "\fB-t\fP <int> (\fIuse <int> threads\fP)"
\&
'in -2m
.ZI 2m "\fB-j\fP <intj> (\fIthis job has index <intj>\fP)"
\&
'in -2m
.ZI 2m "\fB-J\fP <intJ> (\fIa total of <intJ> jobs are used\fP)"
\&
'in -2m
'in +2m
\&
.br
For very large graphs (millions of nodes) and clusterings with large
clusters it may be helpful to allow this program to use multiple CPUs\&.
Additionally it is possible to spread the computation over multiple
jobs/machines\&. These three options are described in the \fBclmprotocols\fP manual page\&.
The following set of options, if given to as many commands, defines three jobs, each running four threads\&.

.di ZV
.in 0
.nf \fC
-t 4 -J 3 -j 0 -o out\&.0
-t 4 -J 3 -j 1 -o out\&.1
-t 4 -J 3 -j 2 -o out\&.2
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

The output can then be collected with

.di ZV
.in 0
.nf \fC
clxdo add_table out\&.[0-2]
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

.in -2m
.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.
.SH REFERENCES

[1] Stijn van Dongen\&. \fIPerformance criteria for graph clustering and Markov
cluster experiments\fP\&. Technical Report INS-R0012, National Research
Institute for Mathematics and Computer Science in the Netherlands,
Amsterdam, May 2000\&.
.br
http://www\&.cwi\&.nl/ftp/CWIreports/INS/INS-R0012\&.ps\&.Z


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shcl/clminfo2.c
/*   (C) Copyright 2006, 2007, 2008, 2009, 2010, 2011, 2012  Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/


/* NOTE:
 * with groups --list should be implied ...
*/

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <time.h>

#include "clm.h"

#include "util/types.h"
#include "util/io.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/compile.h"

#include "impala/io.h"
#include "impala/matrix.h"
#include "impala/tab.h"
#include "impala/stream.h"
#include "impala/app.h"

#include "mcl/interpret.h"
#include "mcl/transform.h"

#include "clew/claw.h"
#include "clew/scan.h"
#include "clew/clm.h"
#include "clew/cat.h"

#include "gryphon/path.h"


enum
{  MY_OPT_OUTPUT = CLM_DISP_UNUSED
,  MY_OPT_LIST
,  MY_OPT_PI
,  MY_OPT_TF
,  MY_OPT_THREAD
,  MY_OPT_G
,  MY_OPT_littleG
,  MY_OPT_CLTREE
,  MY_OPT_CLCEIL
,  MY_OPT_NCLMAX
}  ;

static const char* me  =  "clminfo";

static mcxOptAnchor info2Options[] =
{  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_OUTPUT
   ,  "<fname>"
   ,  "output file name"
   }
,  {  "--list"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_LIST
   ,  NULL
   ,  "list efficiency for all nodes"
   }
,  {  "-t"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_THREAD
   ,  "<num>"
   ,  "number of threads to use"
   }
,  {  "-J"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_G
   ,  "<int>"
   ,  "number of compute jobs overall"
   }
,  {  "-j"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_littleG
   ,  "<int>"
   ,  "index of this compute job"
   }
,  {  "-pi"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_PI
   ,  "<num>"
   ,  "apply inflation with parameter <num> beforehand"
   }
,  {  "-tf"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_TF
   ,  "<tf-spec>"
   ,  "first apply tf-spec to matrix"
   }
,  {  "-cl-tree"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CLTREE
   ,  "<fname>"
   ,  "assume mclcm-type hierarchical clusterings (stack format)"
   }
,  {  "-cat-max"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_NCLMAX
   ,  "<num>"
   ,  "do at most <num> levels of the hierarchy"
   }
,  {  "-cl-ceil"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CLCEIL
   ,  "<num>"
   ,  "skip clusters with <num> or more members"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


static mcxIO* xfstack_g    =  (void*) -1;
static  dim progress_g     =  0;
static  dim n_thread_l     =  -1;
static dim n_group_G       =  -1;
static dim i_group_g       =  -1;
static const char* out_g   =  (void*) -1;

static dim     clceil_g    =  -1;
static double inflation_g  =  -1;
static mcxbool list_g      =  -1;
static mcxIO*  xfout_g     =  (void*) -1;
static mcxTing* tfting_g   =  (void*) -1;
static dim     n_cl_max_g  =  -1;


static mcxstatus info2Init
(  void
)
   {  progress_g  =  0
   ;  xfstack_g   =  NULL
   ;  out_g       =  "-"
   ;  n_thread_l  =  1
   ;  n_group_G   =  1
   ;  i_group_g     =  0
   ;  clceil_g    =  100000
   ;  inflation_g =  -1.0
   ;  tfting_g         =  NULL
   ;  list_g    =  FALSE
   ;  xfout_g     =  mcxIOnew("-", "w")
   ;  n_cl_max_g       =  0
   ;  return STATUS_OK
;  }


static mcxstatus info2ArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case MY_OPT_CLTREE
      :  xfstack_g = mcxIOnew(val, "r")
      ;  break
      ;

         case MY_OPT_OUTPUT
      :  mcxIOrenew(xfout_g, val, "w")
      ;  break
      ;

         case MY_OPT_THREAD
      :  n_thread_l = atoi(val)
      ;  break
      ;

         case MY_OPT_littleG
      :  i_group_g =  atoi(val)
      ;  break
      ;

         case MY_OPT_G
      :  n_group_G =  atoi(val)
      ;  break
      ;

         case MY_OPT_TF
      :  tfting_g = mcxTingNew(val)
      ;  break
      ;

         case MY_OPT_CLCEIL
      :  clceil_g = atoi(val)
      ;  break
      ;

         case MY_OPT_LIST
      :  list_g = TRUE
      ;  break
      ;

         case MY_OPT_NCLMAX
      :  n_cl_max_g = atoi(val)
      ;  break
      ;

         case MY_OPT_PI
      :  inflation_g = atof (val)
      ;  break
      ;

         default
      :  mcxExit(1) 
      ;
      }

   ;  return STATUS_OK
;  }



struct info_bundle
{  mclv* tabulator
;  mclx* cl
;  mclx* cltp
;  dim   clmaxsize
;
}  ;


double info_do
(  const mclx* mx
,  const mclv* cl
,  dim idx
)
   {  double eff, eff_max
   ;  clmVScore sc
   ;  clmVScanDomain(mx->cols+idx, cl, &sc)
   ;  clmVScoreCoverage(&sc, &eff, &eff_max)
   ;  return eff
;  }


#define INCLUDE_NODE(thecls, ib)  (thecls->n_ivps > 0 && ib.cl->cols[thecls->ivps[0].idx].n_ivps <= ib.clmaxsize)


static void info_dispatch
(  mclx* mx
,  dim nid
,  void* data
,  dim thread_id
)
   {  struct info_bundle* ib = data
   ;  const mclv* thecls = ib->cltp->cols+nid
   ;  if (INCLUDE_NODE(thecls, ib[0]))
      ib->tabulator->ivps[nid].val  = info_do(mx, ib->cl->cols+thecls->ivps[0].idx, nid)
   ;  else
      ib->tabulator->ivps[nid].val  = 0.0
;  }


static mcxstatus info2Main
(  int          argc
,  const char*  argv[]
)
   {  mclx* mx
   ;  struct info_bundle ib = { NULL, NULL }
   ;  int a =  0

   ;  mcxIO* xfmx = mcxIOnew(argv[a++], "r")
   ;  mx = mclxReadx(xfmx, EXIT_ON_FAIL, MCLX_REQUIRE_GRAPH)
   ;  mcxIOfree(&xfmx)
   ;  mclxAdjustLoops(mx, mclxLoopCBmax, NULL)
   ;  if (inflation_g >= 0)
      mclxInflate(mx, inflation_g)
   ;  if (tfting_g)
      {  mclgTF* tfar = mclgTFparse(NULL, tfting_g)
      ;  if (!tfar)
         mcxDie(1, me, "errors in tf-spec")
      ;  mclgTFexec(mx, tfar)
   ;  }

      ib.tabulator = mclvCopy(NULL, mx->dom_cols)

   ;  if (!clceil_g)
      clceil_g = N_COLS(mx)

   ;  mcxLogLevel = MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_GAUGE | MCX_LOG_WARN

   ;  mclx_app_init(stderr)
   ;  mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_YES)

   ;  mcxIOopen(xfout_g, EXIT_ON_FAIL)

   ;  n_thread_l = mclx_set_threads_or_die("clm info2", n_thread_l, i_group_g, n_group_G)

   ;  if (n_group_G > 1)
      list_g = TRUE

   ;  while(a < argc)
      {  mcxIO *xfcl
      ;  dim j
      ;  mclxCat st
      ;  mcxstatus status

      ;  if (!strcmp(argv[a], "--"))
         {  a++
         ;  fputc('\n', xfout_g->fp)
         ;  continue
      ;  }

         xfcl =  mcxIOnew(argv[a], "r")

      ;  mclxCatInit(&st)

      ;  if
         (( status
         =  mclxCatRead
            (  xfcl, &st, n_cl_max_g, NULL, mx->dom_rows
            ,  MCLX_PRODUCE_DOMSTACK | MCLX_REQUIRE_PARTITION
            )
         ))
         mcxDie(1, me, "error reading stack")

      ;  for (j=0;j<st.n_level;j++)
         {  ib.cl = st.level[j].mx
         ;  ib.cltp = mclxTranspose(ib.cl)
         ;  ib.clmaxsize = clceil_g
         ;  double sum = 0.0
         ;  dim n_included = 0
         ;  int i

         ;  mclvMakeConstant(ib.tabulator, 0.0)

         ;  if (n_thread_l * n_group_G > 1)
            mclxVectorDispatchGroup(mx, &ib, n_thread_l, info_dispatch, n_group_G, i_group_g, NULL)
         ;  else
            {  for (i=0;i<N_COLS(ib.cltp);i++)           /* fixme: this requires canonical domains */
               {  mclv* thecls = ib.cltp->cols+i
               ;  if (INCLUDE_NODE(thecls, ib))
                  ib.tabulator->ivps[i].val = info_do(mx, ib.cl->cols+thecls->ivps[0].idx, i)
            ;  }
            }

            for (i=0;i<N_COLS(ib.cltp);i++)
            {  double val = ib.tabulator->ivps[i].val
            ;  mclv* thecls = ib.cltp->cols+i
            ;  if (INCLUDE_NODE(thecls, ib))
               n_included++
            ;  if (list_g)
               {  if (i) fputc('\t', xfout_g->fp)
               ;  fprintf(xfout_g->fp, "%.4f", val)
            ;  }
               sum += val
         ;  }
            if (!list_g)
            fprintf(xfout_g->fp, "%.4f", n_included ? sum * 1.0 / n_included : 0.0)
         ;  fputc('\n', xfout_g->fp)
         ;  mclxFree(&ib.cltp)
      ;  }

         mcxIOfree(&xfcl)         /* todo: free stack */
      ;  a++
   ;  }

      mclxFree(&mx)
   ;  return STATUS_OK
;  }



mcxDispHook* mcxDispHookInfo2
(  void
)
   {  static mcxDispHook info2Entry
   =  {  "info2"
      ,  "info2 [options] <mx-file> <cl-file>+"
      ,  info2Options
      ,  sizeof(info2Options)/sizeof(mcxOptAnchor) - 1

      ,  info2ArgHandle
      ,  info2Init
      ,  info2Main

      ,  1
      ,  -1
      ,  MCX_DISP_DEFAULT
      }
   ;  return &info2Entry
;  }


